File: android/app/src/main/kotlin/com/olivier/ettlin/geomessage/service/localisationService.kt
package com.olivier.ettlin.geomessage.service

import android.content.Context
import android.location.Location
import android.os.Looper
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.*
import com.olivier.ettlin.geomessage.model.Message
import com.google.android.gms.tasks.Task
import android.Manifest
import android.content.pm.PackageManager

class LocalisationService(private val context: Context) {

    private val fusedLocationClient: FusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)

    // Vérifie si le service de localisation est activé
    fun isLocationServiceEnabled(): Boolean {
        val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as android.location.LocationManager
        return locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER)
    }

    // Fonction pour obtenir la localisation actuelle
    fun getCurrentLocation(callback: (Location) -> Unit) {
        if (!isLocationServiceEnabled()) {
            Toast.makeText(context, "Le service de localisation est désactivé.", Toast.LENGTH_SHORT).show()
            return
        }

        // Vérification des permissions de localisation
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
            ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // Demander les permissions si elles ne sont pas accordées
            return
        }

        // Récupérer la position actuelle
        fusedLocationClient.lastLocation.addOnCompleteListener { task: Task<Location?> ->
            val location = task.result
            if (location != null) {
                callback(location)
            } else {
                fusedLocationClient.requestLocationUpdates(
                    LocationRequest.create().apply {
                        interval = 10000
                        fastestInterval = 5000
                        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
                    },
                    object : LocationCallback() {
                        override fun onLocationResult(locationResult: LocationResult) { // Utilisez directement LocationResult ici
                            if (locationResult.locations.isNotEmpty()) {
                                callback(locationResult.locations[0])
                                fusedLocationClient.removeLocationUpdates(this)
                            }
                        }
                    },
                    Looper.getMainLooper()
                )
            }
        }
    }

    // Vérifie si la localisation actuelle est dans le rayon du message
    fun checkCurrentLocationInRadius(message: Message, callback: (Boolean) -> Unit) {
        if (!isLocationServiceEnabled()) {
            //Toast.makeText(context, "GPS désactivé", Toast.LENGTH_SHORT).show()
            println("gps désactivé")
            callback(false)
            return
        }

        getCurrentLocation { location ->
            val distance = FloatArray(1)
            Location.distanceBetween(
                location.latitude,
                location.longitude,
                message.latitude,
                message.longitude,
                distance
            )

            val isInRadius = distance[0] <= message.radius
            println("Distance actuelle : ${distance[0]} mètres, Rayon autorisé : ${message.radius} mètres")
            callback(isInRadius)
        }
    }
}

File: lib/services/localisationService.dart

import 'package:geolocator/geolocator.dart';
import 'package:geomessage/model/message.dart';
import 'package:latlong2/latlong.dart';

class LocalisationService {

  Future<LatLng> getCurrentLocation() async {
    bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
    LatLng paris = LatLng(48.866667, 2.333333);
    if (!serviceEnabled) {
       print("Le service de localisation est désactivé.");
      return paris;
    }

    LocationPermission permission = await Geolocator.checkPermission();

    if (permission == LocationPermission.denied) {
      permission = await Geolocator.requestPermission();
      if (permission == LocationPermission.denied) {
        print("Permission de localisation refusée");
        return paris;
      }
    }

    if (permission == LocationPermission.deniedForever) {
      // Si la permission est définitivement refusée, vous pouvez envoyer l'utilisateur aux paramètres pour modifier les permissions
      print("Permission de localisation refusée définitivement");
      return paris;
    }

    // Récupère la position actuelle
    Position position = await Geolocator.getCurrentPosition();

    return LatLng(position.latitude, position.longitude);
  }

  Future<bool> checkGps() async {
    return await Geolocator.isLocationServiceEnabled();
  }

  Future<bool> checkCurrentLocationInRadius(Message message) async {
    bool isGpsEnabled = await checkGps();
    if (!isGpsEnabled) {
      print("GPS désactivé");
      return false;
    }

    Position position = await Geolocator.getCurrentPosition(
      desiredAccuracy: LocationAccuracy.high,
    );

    double distance = Geolocator.distanceBetween(
      position.latitude,
      position.longitude,
      message.latitude,
      message.longitude,
    );

    print("Distance actuelle : $distance mètres, Rayon autorisé : ${message.radius} mètres");

    return distance <= message.radius;
  }


}